File: src/services/liveActivityService.ts
import { Platform } from 'react-native';
import type { WorkoutSession, SessionExercise } from '@/types';

// Lazy load the native module to avoid crashes when not available
let LiveActivityModule: typeof import('expo-live-activity') | undefined;
let moduleLoadAttempted = false;
let currentActivityId: string | undefined;

function getLiveActivityModule() {
  if (!moduleLoadAttempted && Platform.OS === 'ios') {
    moduleLoadAttempted = true;
    try {
      LiveActivityModule = require('expo-live-activity');
    } catch (e) {
      // Module not available - silently continue
    }
  }
  return LiveActivityModule;
}

/**
 * Check if Live Activities are available on this device
 * Requires iOS 16.2+ and a development build
 */
export function isLiveActivityAvailable(): boolean {
  if (Platform.OS !== 'ios') return false;
  try {
    const module = getLiveActivityModule();
    return module !== undefined;
  } catch {
    return false;
  }
}

/**
 * Format elapsed time as MM:SS or HH:MM:SS
 */
function formatElapsedTime(startTime: string): string {
  const start = new Date(startTime).getTime();
  const now = Date.now();
  const elapsed = Math.floor((now - start) / 1000);

  const hours = Math.floor(elapsed / 3600);
  const minutes = Math.floor((elapsed % 3600) / 60);
  const seconds = elapsed % 60;

  if (hours > 0) {
    return `${hours}:${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`;
  }
  return `${minutes}:${seconds.toString().padStart(2, '0')}`;
}

/**
 * Format weight with unit
 */
function formatWeight(weight: number | undefined, unit: string | undefined): string {
  if (!weight) return 'BW';
  return `${weight} ${unit || 'lbs'}`;
}

/**
 * Build the display state for an active set
 */
function buildActiveSetState(
  session: WorkoutSession,
  exercise: SessionExercise,
  setIndex: number,
  progress: { completed: number; total: number }
): { title: string; subtitle: string } {
  const set = exercise.sets[setIndex];
  const setNumber = setIndex + 1;
  const totalSets = exercise.sets.length;

  const weight = formatWeight(set.targetWeight, set.targetWeightUnit);
  const reps = set.targetReps || '?';

  return {
    title: exercise.exerciseName,
    subtitle: `Set ${setNumber}/${totalSets} \u2022 ${weight} \u00D7 ${reps}`,
  };
}

/**
 * Build the display state for rest period
 */
function buildRestState(
  restSeconds: number,
  nextExercise: SessionExercise | null
): { title: string; subtitle: string; timerEndDate: number } {
  const timerEndDate = Date.now() + restSeconds * 1000;
  const nextPreview = nextExercise ? `Next: ${nextExercise.exerciseName}` : 'Finishing up';

  return {
    title: 'Rest',
    subtitle: nextPreview,
    timerEndDate,
  };
}

/**
 * Start a Live Activity for a workout session
 */
export function startWorkoutLiveActivity(
  session: WorkoutSession,
  exercise: SessionExercise | null,
  setIndex: number,
  progress: { completed: number; total: number }
): void {
  if (!isLiveActivityAvailable()) {
    return;
  }

  const module = getLiveActivityModule();
  if (!module) {
    return;
  }

  try {
    // End any existing activity first
    if (currentActivityId) {
      endWorkoutLiveActivity();
    }

    const state = exercise
      ? buildActiveSetState(session, exercise, setIndex, progress)
      : { title: session.name, subtitle: 'Starting workout...' };

    const contentState = {
      title: state.title,
      subtitle: state.subtitle,
      progressBar: { progress: progress.total > 0 ? progress.completed / progress.total : 0 },
    };

    const presentation = {
      backgroundColor: '#1a1a1a',
      titleColor: '#ffffff',
      subtitleColor: '#a0a0a0',
      progressViewTint: '#4CAF50',
    };

    const activityId = module.startActivity(contentState, presentation);

    if (activityId) {
      currentActivityId = activityId;
    }
  } catch (error) {
    // Silently fail - Live Activities are optional
  }
}

/**
 * Update the Live Activity with current workout state
 */
export function updateWorkoutLiveActivity(
  session: WorkoutSession,
  exercise: SessionExercise | null,
  setIndex: number,
  progress: { completed: number; total: number },
  restTimer?: { remainingSeconds: number; nextExercise: SessionExercise | null }
): void {
  if (!isLiveActivityAvailable()) {
    return;
  }

  if (!currentActivityId) {
    return;
  }

  const module = getLiveActivityModule();
  if (!module) {
    return;
  }

  try {
    let updatePayload: any;

    if (restTimer && restTimer.remainingSeconds > 0) {
      // Rest state - show countdown
      const restState = buildRestState(restTimer.remainingSeconds, restTimer.nextExercise);
      updatePayload = {
        title: restState.title,
        subtitle: restState.subtitle,
        progressBar: { date: restState.timerEndDate },
      };
    } else if (exercise) {
      // Active set state
      const state = buildActiveSetState(session, exercise, setIndex, progress);
      updatePayload = {
        title: state.title,
        subtitle: state.subtitle,
        progressBar: { progress: progress.total > 0 ? progress.completed / progress.total : 0 },
      };
    } else {
      return;
    }

    module.updateActivity(currentActivityId, updatePayload);
  } catch (error) {
    // Silently fail - Live Activities are optional
  }
}

/**
 * End the Live Activity with a completion message
 */
export function endWorkoutLiveActivity(message?: string): void {
  if (!isLiveActivityAvailable()) {
    return;
  }

  if (!currentActivityId) {
    return;
  }

  const module = getLiveActivityModule();
  if (!module) {
    return;
  }

  try {
    const endPayload = {
      title: message || 'Workout Complete',
      subtitle: 'Great job!',
    };

    module.stopActivity(currentActivityId, endPayload);
    currentActivityId = undefined;
  } catch (error) {
    // Silently fail - Live Activities are optional
  }
}

File: swift-ios/LiftMark/Services/LiveActivityService.swift
import Foundation

#if os(iOS)
import ActivityKit

// MARK: - Workout Activity Attributes

struct WorkoutActivityAttributes: ActivityAttributes {
    public struct ContentState: Codable, Hashable {
        var title: String
        var subtitle: String
        var progress: Double // 0.0 to 1.0
        var timerEndDate: Date?
    }

    var workoutName: String
}
#endif

// MARK: - LiveActivityService

final class LiveActivityService {
    static let shared = LiveActivityService()

    #if os(iOS)
    @available(iOS 16.2, *)
    private var currentActivity: Activity<WorkoutActivityAttributes>?
    #endif

    private init() {}

    // MARK: - Availability

    /// Check if Live Activities are available (iOS 16.2+).
    func isAvailable() -> Bool {
        #if os(iOS)
        if #available(iOS 16.2, *) {
            return ActivityAuthorizationInfo().areActivitiesEnabled
        }
        #endif
        return false
    }

    // MARK: - Start

    /// Start a Live Activity for a workout session.
    func startWorkoutActivity(
        session: WorkoutSession,
        exercise: SessionExercise?,
        setIndex: Int,
        progress: (completed: Int, total: Int)
    ) {
        #if os(iOS)
        guard isAvailable() else { return }

        if #available(iOS 16.2, *) {
            // End any existing activity first
            endWorkoutActivity()

            do {
                let state: WorkoutActivityAttributes.ContentState
                if let exercise {
                    state = buildActiveSetState(exercise: exercise, setIndex: setIndex, progress: progress)
                } else {
                    state = WorkoutActivityAttributes.ContentState(
                        title: session.name,
                        subtitle: "Starting workout...",
                        progress: 0
                    )
                }

                let attributes = WorkoutActivityAttributes(workoutName: session.name)
                let content = ActivityContent(state: state, staleDate: nil)

                let activity = try Activity.request(
                    attributes: attributes,
                    content: content,
                    pushType: nil
                )
                currentActivity = activity
            } catch {
                // Silently fail — Live Activities are optional
            }
        }
        #endif
    }

    // MARK: - Update

    /// Update the Live Activity with current workout state.
    func updateWorkoutActivity(
        session: WorkoutSession,
        exercise: SessionExercise?,
        setIndex: Int,
        progress: (completed: Int, total: Int),
        restTimer: (remainingSeconds: Int, nextExercise: SessionExercise?)? = nil
    ) {
        #if os(iOS)
        guard isAvailable() else { return }

        if #available(iOS 16.2, *) {
            guard let activity = currentActivity else { return }

            let state: WorkoutActivityAttributes.ContentState

            if let restTimer, restTimer.remainingSeconds > 0 {
                state = buildRestState(
                    remainingSeconds: restTimer.remainingSeconds,
                    nextExercise: restTimer.nextExercise,
                    progress: progress
                )
            } else if let exercise {
                state = buildActiveSetState(exercise: exercise, setIndex: setIndex, progress: progress)
            } else {
                return
            }

            let content = ActivityContent(state: state, staleDate: nil)
            Task {
                await activity.update(content)
            }
        }
        #endif
    }

    // MARK: - End

    /// End the Live Activity with an optional completion message.
    func endWorkoutActivity(message: String? = nil) {
        #if os(iOS)
        guard isAvailable() else { return }

        if #available(iOS 16.2, *) {
            guard let activity = currentActivity else { return }

            let finalState = WorkoutActivityAttributes.ContentState(
                title: message ?? "Workout Complete",
                subtitle: "Great job!",
                progress: 1.0
            )

            let content = ActivityContent(state: finalState, staleDate: nil)
            Task {
                await activity.end(content, dismissalPolicy: .default)
            }
            currentActivity = nil
        }
        #endif
    }

    // MARK: - Display State Builders

    #if os(iOS)
    @available(iOS 16.2, *)
    private func buildActiveSetState(
        exercise: SessionExercise,
        setIndex: Int,
        progress: (completed: Int, total: Int)
    ) -> WorkoutActivityAttributes.ContentState {
        let set = exercise.sets.indices.contains(setIndex) ? exercise.sets[setIndex] : nil
        let setNumber = setIndex + 1
        let totalSets = exercise.sets.count

        let weight = formatWeight(set?.targetWeight, unit: set?.targetWeightUnit)
        let reps = set?.targetReps.map { String($0) } ?? "?"

        let progressValue = progress.total > 0 ? Double(progress.completed) / Double(progress.total) : 0

        return WorkoutActivityAttributes.ContentState(
            title: exercise.exerciseName,
            subtitle: "Set \(setNumber)/\(totalSets) \u{2022} \(weight) \u{00D7} \(reps)",
            progress: progressValue
        )
    }

    @available(iOS 16.2, *)
    private func buildRestState(
        remainingSeconds: Int,
        nextExercise: SessionExercise?,
        progress: (completed: Int, total: Int)
    ) -> WorkoutActivityAttributes.ContentState {
        let nextPreview = nextExercise.map { "Next: \($0.exerciseName)" } ?? "Finishing up"
        let progressValue = progress.total > 0 ? Double(progress.completed) / Double(progress.total) : 0
        let timerEnd = Date().addingTimeInterval(TimeInterval(remainingSeconds))

        return WorkoutActivityAttributes.ContentState(
            title: "Rest",
            subtitle: nextPreview,
            progress: progressValue,
            timerEndDate: timerEnd
        )
    }
    #endif

    private func formatWeight(_ weight: Double?, unit: WeightUnit?) -> String {
        guard let weight, weight > 0 else { return "BW" }
        return "\(Int(weight)) \(unit?.rawValue ?? "lbs")"
    }
}
